docs(devs-start): about web components (#1687)
RedHat-UX/red-hat-design-system: docs/get-started/developers/index.md

Reviewer: there's some more prose https://github.com/RedHat-UX/red-hat-design-system/wiki#custom-elements that you are welcome to crib

  collection: sortedDevelopers
  order: 00
order: 20
importElements:
  - rh-card
  - rh-cta
---

<style>
  .grid > rh-card {
    display: grid;
  }

  rh-card > svg[slot] {
    fill: var(--rh-color-accent-brand-on-light, #ee0000);
    width: var(--rh-size-icon-07, 80px);
    margin-block-end: var(--rh-space-lg, 16px);
  }

  rh-card > p[slot],
  rh-card > h3[slot] {
    margin-block: 0;
  }

  #section-encapsulation,
  #section-apis,
  #section-framework {

>     display: grid;

Reviewer 2: Instead of defining this grid here, would it be possible to use or expand upon the [grid classes](https://github.com/RedHat-UX/red-hat-design-system/blob/main/docs/styles/grid.css) Steve uses throughout the uxdot site?

Author: Initially, I was hesitant to expand on the grid, but you know... why not?!

Author: @zeroedin would love your input on the "expansion" of `.grid` to handle content + image layouts.

Reviewer 3: As discussed on a video chat but adding my response here, these styles would be best moved to a `layouts` layer and "de-BEMified".  

The intent with the grid stylesheet was to act as a shareable artifact ie (reset, fonts, typography, base, grid) and most likely it will be replaced by some other shared style or component such as `rhx-grid` or whatever.  

Ultimately these layout styles could also inform a layout pattern that becomes a shared artifact and less ux.redhat.com specific,  IMHO they would be best served decoupled from each other.

Author: Thanks, @zeroedin. We'll tackle this in another PR that refactors the `layouts` and `grid` we've got in ux dot with `rhx-grid`

## About web components

> Web components are based on a set of web platform APIs that help to create reusable and encapsulated UI elements. Those standards consist of custom elements, shadow DOM, and HTML Templates.

Reviewer: there as to be phrasing content between headings, iirc. checking with @adamjohnson

Reviewer 2: Greg answered this [in Slack](https://redhat-internal.slack.com/archives/C05GTN4E82G/p1719930122638249), but:

✔️ This works:
```html
<h1>Heading 1</h1>
<h2>Heading 2</h2>
...content...
```

❌  This does not:
```html
<h2>Heading 2</h2>
<h2>Heading 2</h2>
```

❌  Nor this:
```html
<h2>Heading 2</h2>
...content...
<h4>Heading 4</h4>
```

Reviewer: Thanks for clarifying 😊

> Web components are based on a set of web platform APIs that help to create reusable and encapsulated UI elements. Those standards consist of [custom elements](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements), [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM), and [HTML Templates](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_templates_and_slots).

Reviewer: ce's don't neccessarily have to be reusable, or encapsulated

but they can be either or both

other standards: css custom properties, css shadow parts, ecmascript modules, element internals

Author: @bennypowers can you suggest a rephrasing of this?

### Why Web Components?

#### Encapsulation

> Web components allow for encapsulation, which provides a clear boundary between the component and the website. This makes the components modular and prevents conflicts with page-level CSS and JS, hiding complex logic from the UI, and allowing for a scalable design system. In addition to encapsulating styles, web components also ship complex accessibility features.

Reviewer: ```suggestion
Web components allow for encapsulation, which provides a clear boundary between the component and the website. This makes the components modular and prevents conflicts with page-level CSS and JS, hiding complex logic from the UI, and allowing for a scalable design system. In addition to encapsulating styles, web components can also ship complex accessibility features, making it easier for page authors to provide accessible rich experiences.
```

#### Web Standard APIs

> Web Components use Web Standard APIs, which means the technology is future-proofed, reliable, and supported by every modern browser.

Reviewer: ```suggestion
Because Web Components use Web Standards, the technology is future-proofed, reliable, and supported by every modern browser.
```

#### Platform and framework agnostic

Because Web Components are built using Web Standards, they are platform and framework agnostic, allowing Red Hat to build and maintain a single Design System library that can be used across the entire enterprise. This makes it easier to switch frameworks if necessary and reduces duplication of efforts.

### Do Web Components affect page performance?

> Much of our JavaScript payload is strings of HTML or CSS, which allows us to define new, custom HTML elements. This compresses well and parses quickly.

Reviewer: We've observed that the entire payload of adding one of our design system elements to a page is less than the react-dom library

shadow dom encapsulation lets browsers optimize rendering in certain scenarios

    <rh-cta slot="footer">
      <a href="/elements">Our elements</a>
    </rh-cta>
  </rh-card>
  <rh-card>
    <svg slot="header" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 36 36">
      <path d="M31 4.38H5a.62.62 0 0 0-.62.62v26a.62.62 0 0 0 .62.62h26a.62.62 0 0 0 .62-.62V5a.62.62 0 0 0-.62-.62Zm-.62 26H5.62V5.62h24.76Z"/>
      <path d="M17.24 21.54a.62.62 0 0 0 .74-.49l1.24-5.84a.64.64 0 0 0-.48-.75.63.63 0 0 0-.74.54l-1.25 5.84a.64.64 0 0 0 .49.7ZM23.12 18 21 20.11a.63.63 0 1 0 .88.89l2.56-2.56a.62.62 0 0 0 0-.89L21.88 15a.63.63 0 1 0-.88.88Zm-9 3a.63.63 0 0 0 .88-.89L12.88 18 15 15.87c.59-.56-.31-1.46-.88-.88l-2.56 2.56a.62.62 0 0 0 0 .89Z"/>
    </svg>
    <h3 slot="header">GitHub repositories</h3>
    <p>Explore our code, roadmaps, and discussions in the 
      <a href="https://github.com/RedHat-UX/red-hat-design-system">Red Hat Design System repo</a> and the 
      <a href="https://github.com/RedHat-UX/red-hat-design-tokens">Red Hat Design Tokens repo</a>.</p>
    <rh-cta slot="footer">
      <a href="https://github.com/RedHat-UX/">View our repos</a>
    </rh-cta>
  </rh-card>
</div>

## About Web Components

### What are Web Components?

> Every web component is a custom HTML element. Web Components are made with standard web platform APIs, and not only with JavaScript frameworks. Those standards include [Custom Elements][ce]; [Shadow DOM][sd]; [HTML templates][te]; [CSS Custom Properties][cssprop] and [Shadow Parts][csspart]; [Element Internals][internals]; [ECMAScript Modules][esm]; and others. These combine to enable developers to write reusable and encapsulated UI elements which work with the browser's in-built component model.

Reviewer 2: Probably better to use an unordered list for:

> Those standards include:
>  * Custom Elements
>  * Shadow Dom...

Reviewer 2: "in-built" or "built-in"? I'd vote for built-in. YMMV.

Author: I think the list will help readability. Good idea.

      <path d="M94.8564 94.8159L41.2356 65.6934" stroke="#A3A3A3" stroke-width="2" stroke-linejoin="bevel" stroke-dasharray="16 8"/>
      <path d="M92.6193 89.3156L84.5011 74.0371" stroke="#A3A3A3" stroke-width="2" stroke-linejoin="bevel" stroke-dasharray="16 8"/>
      <path d="M80.0355 64.6836L79.5202 77.2998L90.7037 71.438L80.0355 64.6836Z" fill="#A3A3A3"/>
      <path d="M93.4873 143.04L61.5168 195.013" stroke="#A3A3A3" stroke-width="2" stroke-linejoin="bevel" stroke-dasharray="16 8"/>
      <path d="M87.874 144.977L72.1801 152.26" stroke="#A3A3A3" stroke-width="2" stroke-linejoin="bevel" stroke-dasharray="16 8"/>
      <path d="M62.5999 156.215L75.1699 157.41L69.9194 145.927L62.5999 156.215Z" fill="#A3A3A3"/>
      <path d="M152.769 175L138.625 234.357" stroke="#A3A3A3" stroke-width="2" stroke-linejoin="bevel" stroke-dasharray="16 8"/>
      <path d="M155.368 180.339L161.316 196.586" stroke="#A3A3A3" stroke-width="2" stroke-linejoin="bevel" stroke-dasharray="16 8"/>
      <path d="M165.293 206.157L155.56 198.113L167.393 193.706L165.293 206.157Z" fill="#A3A3A3"/>
    </g>
  </svg>
  <div>

#### Encapsulation

> Web Components encapsulate their templates, styles, and behaviour. They establish a strong boundary between the component's internals and the rest of the website, letting developers write more modular code while avoiding conflicts with page-level CSS and JS. This hides complex logic and templating from the rest of the page's UI, which helps to ship design systems and applications at scale. In addition to encapsulating styles, Web Components can ship complex accessibility patterns, making it easier for page authors to provide accessible rich experiences.

Reviewer 2: Change behaviour to behavior.

Reviewer 4: I can always tell when @bennypowers has written something. :upside_down_face:

Author: Ah, the good 'ol Canadian spelling 😉 🇨🇦

> Web Components encapsulate their templates, styles, and behaviour. They establish a strong boundary between the component's internals and the rest of the website, letting developers write more modular code while avoiding conflicts with page-level CSS and JS. This hides complex logic and templating from the rest of the page's UI, which helps to ship design systems and applications at scale. In addition to encapsulating styles, Web Components can ship complex accessibility patterns, making it easier for page authors to provide accessible rich experiences.

Reviewer 4: I think it should be "accessible, rich experiences" at the very end of this paragraph.

    </mask>
    <g mask="url(#mask0_63_262)">
      <path d="M106.629 42.6347L155.311 42.6365L179.419 84.1211L154.844 126.147L106.162 126.416L82.0539 84.6602L106.629 42.6347Z" fill="#92c5f9"/>
      <path d="M96.6929 76.4349L121.562 33.9058L170.845 33.9066L195.211 75.8367L170.339 118.371L121.062 118.643L96.6929 76.4349Z" stroke="#151515" stroke-width="6"/>
      <path d="M171.5 144.457H119L121.903 138.457H168.245L171.5 144.457Z" fill="#a3a3a3"/>
      <path d="M157 122.457L119 122.181L122.844 115.456L161.13 115.406L157 122.457Z" fill="#92c5f9"/>
      <path d="M172.5 122.232H157.13L161.15 115.357L168.6 115.356L172.5 122.232Z" fill="#f2f2f2"/>
      <path d="M172.8 30.1408L199.64 75.8496L191.676 75.85L168.9 36.9563L172.8 30.1408Z" fill="#f2f2f2"/>
      <path d="M186.196 25.0089L212.2 70.013L205.4 70.0167L182.627 30.8276L186.196 25.0089Z" fill="#a3a3a3"/>
      <line x1="172.857" y1="35.4855" x2="195.495" y2="73.2152" stroke="#151515" stroke-width="2" stroke-dasharray="8 8"/>
      <line x1="124" y1="119" x2="168" y2="119" stroke="#151515" stroke-width="2" stroke-dasharray="8 8"/>
    </g>
  </svg>
  <div>

#### Web standards APIs

> Because Web Components are build with web standards, the technology is future-proofed, reliable, and supported by every modern browser. As the browser's native component model, Web Components continually benefit from new specifications and features as they are added to the web platform, for example [Scoped Custom Element Registries][scoped].

Reviewer 2: "Because web components are **built** with web standards..."

      <path d="M183.181 222.68L183.428 222.821L183.676 222.68L191.569 218.181L191.821 218.037V217.746V208.749V208.458L191.569 208.314L183.676 203.815L183.428 203.674L183.181 203.815L175.288 208.314L175.036 208.458V208.749V217.746V218.037L175.288 218.181L183.181 222.68Z" stroke="#A3A3A3"/>
      <path d="M166.178 222.68L166.426 222.821L166.673 222.68L174.566 218.181L174.818 218.037V217.746V208.749V208.458L174.566 208.314L166.673 203.815L166.426 203.674L166.178 203.815L158.285 208.314L158.033 208.458V208.749V217.746V218.037L158.285 218.181L166.178 222.68Z" stroke="#A3A3A3"/>
      <path d="M124.839 100.84L154.298 100.84L168.887 125.945L154.016 151.376L124.556 151.539L109.968 126.271L124.839 100.84Z" fill="#92C5F9"/>
      <path d="M122.93 120.592L137.269 96.07L165.687 96.0693L179.734 120.243L165.393 144.768L136.979 144.926L122.93 120.592Z" stroke="#151515" stroke-width="4"/>
    </g>
  </svg>
  <div>

#### Framework agnostic

> Web Components are framework agnostic, meaning they can be used in any javascript framework which outputs HTML. This allows Red Hat to build and maintain a single Design System library that can be used across the entire enterprise. This makes it easier to switch frameworks if necessary, without throwing away all their work. Web Components help to break down silos, increase collaboration, reduce duplication of effort.

Reviewer 2: javascript > JavaScript

> This makes it easier to switch frameworks, if necessary, without...

(add a comma after "frameworks")

> Web Components are framework agnostic, meaning they can be used in any javascript framework which outputs HTML. This allows Red Hat to build and maintain a single Design System library that can be used across the entire enterprise. This makes it easier to switch frameworks if necessary, without throwing away all their work. Web Components help to break down silos, increase collaboration, reduce duplication of effort.

Reviewer 4: Add an "and" before "reduce duplication of effort"
